metricplugin/wiretap.go

```diff
@@ -149,14 +330,181 @@ func (wt *BitswapWireTap) notifyConnEvent(connected bool, conn network.Conn) {
 
 // Connected is called when a connection is opened.
 func (wt *BitswapWireTap) Connected(_ network.Network, conn network.Conn) {
-	log.Debugf("Connection event for PID: %s, Addr: %s\n", conn.RemotePeer(), conn.RemoteMultiaddr())
+	remotePeer := conn.RemotePeer()
+	log.Debugf("Connection event for peer %s, address: %s", remotePeer, conn.RemoteMultiaddr())
+
+	// Track connection/peer connectivity.
+	wt.connManager.lock.Lock()
+	defer wt.connManager.lock.Unlock()
+	if _, ok := wt.connManager.allConnections[conn.ID()]; ok {
+		log.Warnf("Duplicate connection ID %s (Duplicate Connection Event)", conn.ID())
+		// What now?
+	}
+	wt.connManager.allConnections[conn.ID()] = struct{}{}
+
+	peerConn, ok := wt.connManager.peers[remotePeer]
+	if !ok {
+		peerConn = peerConnection{
+			bitswapSender: nil,
+			connections:   map[string]network.Conn{conn.ID(): conn},
+		}
+	} else {
+		if _, ok := peerConn.connections[conn.ID()]; ok {
+			log.Warnf("Duplicate connection ID %s to peer %s", conn.ID(), remotePeer)
+			// What now?
+		}
+		peerConn.connections[conn.ID()] = conn
+	}
+	wt.connManager.peers[remotePeer] = peerConn
+
+	// Connect via Bitswap
+	// Check if we're shutting down
+	select {
+	case <-wt.closing:
+		log.Debugf("Connection event handler: not starting bitswap connector to %s because we're shutting down", conn.ID())
+	default:
+		// Open Bitswap connection in the background
+		wt.wg.Add(1)
+		go wt.connectBitswap(conn.RemotePeer(), time.Duration(1)*time.Second)
+	}
 
 	wt.notifyConnEvent(true, conn)
 }
 
+// Creates a Bitswap sender to a given peer after waiting the specified time.
+// This is supposed to be run as a goroutine and will decrement wt.wg!
+//
+// Creating a Bitswap sender instructs the IPFS node to connect to the given
+// peer, open a Bitswap stream, and creates a message sender on top of that
+// stream.
+// We will check if we're still connected to the peer, to avoid re-opening
+// connections.
+// We also check if any of the connections we have to a peer are non-transient.
+// If we're still connected, we instruct Bitswap to create a new sender.
+// This is racy: The connection could have been closed in the meantime.
+// In that case, Bitswap will reconnect, which can take some time (or fail).
+// We re-check our connectivity to the peer after we created the sender.
+// If we're not connected anymore (for whatever reason), we destroy the sender.
+//
+// It can happen that the connection that triggered the connection event which
+// then triggered us to connect via Bitswap is not the connection through which
+// the Bitswap sender communicates.
+// This is generally not a problem: The Bitswap sender, when instructed to send
+// a message, will retry and reconnect to its peer, using whatever connection is
+// available to that peer.
+func (wt *BitswapWireTap) connectBitswap(remotePeer peer.ID, waitTime time.Duration) {
```

Comment:
Do remote hosts close streams that are not used? Or do we have a BitSwap stream only once in the lifetime of a connection?
If these things are closed it could be a viable alternative to only open the streams on demand when we're actually requesting something (this then delays the request operation of course)

Reply:
That's a good question 🤔 I don't know the answer, maybe we'll figure this out once we let this run on the monitoring node and compare "number of bitswap senders" to "number of bitswap streams" via prometheus. The bitswap senders don't notice when their underlying stream and/or connection is closed, they'll attempt to connect to the peer and open a new stream once we use them to send a message.

Creating the senders on demand could be a viable alternative, we'll have to see how this works!